src/util/include/util/Color.h

```diff
@@ -11,149 +11,160 @@
+        return uint32_t(lhs) != uint32_t(rhs);
     }
-    template <typename T, std::enable_if_t<std::is_signed_v<T>, int> = 0>
-    constexpr explicit operator T() const {
-        return val;
+    constexpr friend bool operator<(LegacyARGBU8Color const& lhs, LegacyARGBU8Color const& rhs) {
+        return uint32_t(lhs) < uint32_t(rhs);
     }
 
-    constexpr auto operator=(Color const& rhs) -> Color& = default;
-    constexpr auto operator=(uint32_t rhs) -> Color& { return *this = Color(rhs); }
-
-    constexpr friend auto operator&(Color lhs, uint32_t rhs) -> uint32_t { return lhs.val & rhs; }
-    constexpr friend auto operator|(Color lhs, uint32_t rhs) -> uint32_t { return lhs.val | rhs; }
-    constexpr friend auto operator^(Color lhs, uint32_t rhs) -> uint32_t { return lhs.val ^ rhs; }
-    constexpr friend auto operator&(uint32_t lhs, Color rhs) -> uint32_t { return lhs & rhs.val; }
-    constexpr friend auto operator|(uint32_t lhs, Color rhs) -> uint32_t { return lhs | rhs.val; }
-    constexpr friend auto operator^(uint32_t lhs, Color rhs) -> uint32_t { return lhs ^ rhs.val; }
-    constexpr friend auto operator>>(Color lhs, uint32_t rhs) -> uint32_t { return lhs.val << rhs; }
-    constexpr friend auto operator<<(Color lhs, uint32_t rhs) -> uint32_t { return lhs.val >> rhs; }
+    constexpr friend bool operator==(LegacyARGBU8Color const& lhs, uint32_t rhs) { return uint32_t(lhs) == rhs; }
+    constexpr friend bool operator!=(LegacyARGBU8Color const& lhs, uint32_t rhs) { return uint32_t(lhs) != rhs; }
+    constexpr friend bool operator<(LegacyARGBU8Color const& lhs, uint32_t rhs) { return uint32_t(lhs) < rhs; }
 
-    constexpr friend bool operator==(Color lhs, Color rhs) { return lhs.val == rhs.val; }
-    constexpr friend bool operator!=(Color lhs, Color rhs) { return lhs.val != rhs.val; }
-    constexpr friend bool operator<(Color lhs, Color rhs) { return lhs.val < rhs.val; }
-    constexpr friend bool operator<=(Color lhs, Color rhs) { return lhs.val <= rhs.val; }
-    constexpr friend bool operator>(Color lhs, Color rhs) { return lhs.val > rhs.val; }
-    constexpr friend bool operator>=(Color lhs, Color rhs) { return lhs.val >= rhs.val; }
-
-    inline friend std::ostream& operator<<(std::ostream& os, Color rhs) { return os << rhs.val; }
+    inline friend std::ostream& operator<<(std::ostream& os, LegacyARGBU8Color rhs) { return os << uint32_t(rhs); }
+    inline friend std::istream& operator>>(std::istream& is, LegacyARGBU8Color& rhs) {
+        uint32_t val{};
+        is >> val;
+        rhs = val;
+        return is;
+    }
 
-    uint32_t val{};
+    // Legacy ordering (argb, not rgba):
+    // Had to use unions, because the compiler didn't like the constexpr
+    // replace this with c++20: std::bit_cast
+    union {
+        uint32_t val{};
+        struct {
+            uint8_t b;
+            uint8_t g;
+            uint8_t r;
+            uint8_t a;
```

Comment:
It's a matter of endianness: https://en.wikipedia.org/wiki/Endianness
[For reference, a god bold code snipped.](https://godbolt.org/#z:OYLghAFBqd5QCxAYwPYBMCmBRdBLAF1QCcAaPECAMzwBtMA7AQwFtMQByARg9KtQYEAysib0QXACx8BBAKoBnTAAUAHpwAMvAFYTStJg1DIApACYAQuYukl9ZATwDKjdAGFUtAK4sGe1wAyeAyYAHI%2BAEaYxCAAzFykAA6oCoRODB7evnrJqY4CQSHhLFEx8baY9vkMQgRMxASZPn4JdpgO6bX1BIVhkdFxrXUNTdlD3b3FpYMAlLaoXsTI7BzmscHI3lgA1CaxbsgKBPiCe9gmGgCCF5dHxF4O25kkJgDsVlfbX9teDOlvH0u32BP2CBFiZgA%2BgRtgA3MR7QEg753B4EAE3ZHIrxggAc0O2EURmKxwJxgnxMOAxM%2BpO%2B5IIlO2xBpQLpXwZTKYrNJbwAIjzvvzWcLYoCbjcmF4iNs0LQSABJQQQGYY2lfZ7EWWCr7IAB08Nou1ifO2GlUVEw6GQESYAE5cTrmZgCIsGNqxRLXgKrpLpahZZ4SG4EPUVWq2Rqg1rTJ71bK9UxjabzZanfqtXsU6prem9cBk2bVLa8xFC%2BaHU7iC63R7xd6JVcOHNaJwAKy8PwcLSkVCcNzWazbBQLJaYXZmeK8AiaZtzADWIDbGn0nEkndnvc4vAUIBXM%2B7zdIcFgSDQLESdGi5Eo58v9BiwC4XDMfDoBGiu4gEU3EWC9QAT04HhSHPNhBAAeQYWggMPUgsBYQxgHEOD8GrDpYUwXc4MwVR2mlFYQLBSpN1oPAImIQCPCwTcCGIPAWGAo8qAMYAFAANTwTAAHcIMSRgmJkQQRDEdgpCE%2BQlDUTddASAwjBQQdLH0cjd0gOZUESapsIAWiOdAs1MSxrC4V5th0iDYh3Sp2mqFwGHcTxmgkV9AmCPoSgGAA2JIUjSARRhaV9cn8hhJn6GIfLaDoBC6EYnLGV9ouqOKencqZvNsYZGgSoKsomdKIpALy5hHRZlj0OjMEIo9Ww4DtSC7Hs%2Bw4VRcS8nSvMkbZgGQZBtmfPUzG2CAB2M5TtlwQgSAnconlQC8r0zScuBmadZxmBclxXOr10azcWp3PdSAPLRNtXDgzA3ODDpOja5kw4hUmcSQgA%3D%3D)

Reply:
:pray:  Thanks for the explanation.
This means that this code only works on little endian architecture? Or does the compiler adjust depending on the platform?

Reply:
Checked this via godbolt.org on arch, but this could be implementation defined. Removed it completely. Actually, we might have to introduce a byte swapping, depending on the architecture.